### ISSUE TITLE
[request-for-help-with-developing-plugin] Help with module lazy evaluation plugin

### ISSUE THREAD

**Issue by fabiomcosta** [webpack/webpack#9741]
Hello folks,

I've tried to get help/pointer to a Webpack Plugin that would allow lazy evaluation of JS modules on Stackoverflow with no success, so I decided to post this here, in the hope that it would get more visibility.

The SO post explains what I mean by lazy evaluation.
https://stackoverflow.com/questions/57262947/lazy-evaluation-of-modules-with-webpack?noredirect=1#comment101152378_57262947

If no such plugin already exists but you have any pointers on how I could achieve that via a plugin, please let me know as well.

From the example on the SO link, the following code:
```js
import _ from 'lodash';

export function toUpperCase(str) {
  return _.toUpper(str);
}
```

Should become:
```js
module.exports.toUpperCase = function(str) {
  return require('lodash').default.toUpper(str);
}
```

I think this could be achieved by tapping on the `import` Parser hook to remove the module level import statement, but I'm not sure how I can change the imported module identifier to the `require` call shown on the example.

Any pointers or tips on how to achieve this, or to plugins that do this or something similar would be greatly appreciated.

**Comment by evilebottnawi** [comment]
Better use babel for this purpose, not webpack

**Comment by fabiomcosta** [comment]
@evilebottnawi thank you for reading and answering!
100% agreed, Babel even has this open available via its `@babel/plugin-transform-modules-commonjs` plugin, but we can't use that because optimizations like tree shaking and module concatenation won't work properly because WP runs after Babel and won't be able to "see" the module on its ESM form. This is the same reason why we set "modules" to false on Babel's preset-env.
I explained this further on the SO link: https://stackoverflow.com/questions/57262947/lazy-evaluation-of-modules-with-webpack?noredirect=1#comment101152378_57262947
